src/tree/gpu_hist/gradient_based_sampler.cu

```diff
@@ -0,0 +1,286 @@
+  if (is_external_memory_) {
+    thrust::transform(dh::tbegin(gpair), dh::tend(gpair),
+                      thrust::counting_iterator<size_t>(0),
+                      dh::tbegin(row_weight_),
+                      CalculateWeight(rnd));
+    return SequentialPoissonSampling(gpair, dmat);
+  } else {
+    float normalization = thrust::transform_reduce(dh::tbegin(gpair), dh::tend(gpair),
+                                                   CombineGradientPair(),
+                                                   0.0f,
+                                                   thrust::plus<float>());
+    // Set gradient pair to 0 with p = 1 - p_i
+    thrust::transform(dh::tbegin(gpair), dh::tend(gpair),
+                      thrust::counting_iterator<size_t>(0),
+                      dh::tbegin(gpair),
+                      PoissonSampling(sample_rows_, normalization, rnd));
+    return {dmat->Info().num_row_, original_page_, gpair};
+  }
+}
+
+/*! \brief A functor that returns true if the gradient pair is non-zero. */
+struct IsNonZero : public thrust::unary_function<GradientPair, bool> {
+  XGBOOST_DEVICE bool operator()(const GradientPair& gpair) const {
+    return gpair.GetGrad() != 0 || gpair.GetHess() != 0;
+  }
+};
+
+/*! \brief A functor that clears the row indexes with empty gradient. */
+struct ClearEmptyRows : public thrust::binary_function<GradientPair, size_t, size_t> {
+  XGBOOST_DEVICE size_t operator()(const GradientPair& gpair, size_t row_index) const {
+    if (gpair.GetGrad() != 0 || gpair.GetHess() != 0) {
+      return row_index;
+    } else {
+      return SIZE_MAX;
+    }
+  }
+};
+
+// Perform sampling after the weights are calculated.
+GradientBasedSample GradientBasedSampler::SequentialPoissonSampling(
+      common::Span<GradientPair> gpair, DMatrix* dmat) {
+  // Sort the gradient pairs and row indexes by weight.
+  thrust::sort_by_key(dh::tbegin(row_weight_), dh::tend(row_weight_),
+                      thrust::make_zip_iterator(thrust::make_tuple(dh::tbegin(gpair),
+                                                                   dh::tbegin(row_index_))));
+
+  // Clear the gradient pairs not included in the sample.
+  thrust::fill(dh::tbegin(gpair) + sample_rows_, dh::tend(gpair), GradientPair());
+
+  // Mask the sample rows.
+  thrust::fill(dh::tbegin(sample_row_index_), dh::tbegin(sample_row_index_) + sample_rows_, 1);
+  thrust::fill(dh::tbegin(sample_row_index_) + sample_rows_, dh::tend(sample_row_index_), 0);
+
+  // Sort the gradient pairs and sample row indexes by the original row index.
+  thrust::sort_by_key(dh::tbegin(row_index_), dh::tend(row_index_),
+                      thrust::make_zip_iterator(thrust::make_tuple(dh::tbegin(gpair),
+                                                                   dh::tbegin(sample_row_index_))));
+
+  // Compact the non-zero gradient pairs.
+  thrust::copy_if(dh::tbegin(gpair), dh::tend(gpair), dh::tbegin(gpair_), IsNonZero());
```

Comment:
Is it possible that less values are copied than sample_rows_?

Reply:
Yes it's possible. Added a line to fill `gpair_` with empty gpairs.